crates/prices/src/core/swap.rs

```diff
@@ -0,0 +1,83 @@
+use super::eth;
+
+/// A token swap. Specifies how much of one token should be converted to another
+/// token.
+#[derive(Debug, Clone, Copy)]
+pub struct Swap {
+    /// The token to swap from.
+    pub from: FromToken,
+    /// The token to swap into.
+    pub to: ToToken,
+    /// The amount to swap.
+    pub amount: FromAmount,
+}
```

Comment:
This abstraction reads like it's only intended for sell orders.
For buy orders you only know how much `ToToken` you want to buy and the price estimator is supposed to tell you how much `FromToken` you need for that. That means you couldn't really populate the `amount` field with a reasonable value for buy orders.
Please correct me if I misunderstood this struct.

Reply:
You're right, this is an oversight. For some reason I thought this covers both buy and sell orders.

Reply:
The result of a `Swap` are an estimated amount and fee, both expressed in the `to` token.

For an order which wants to _sell_ 100 token A for token B, the swap would look like this:
```
Swap { from: A, to: B, amount: 100 }
```
The fee would be denominated in token B (our buy token). However, since we know the conversion rate between A (our sell token) and B, we can convert that fee into token A. (Note: since we're working with estimates these kinds of conversions are completely acceptable.) $\Box$

For an order which wants to _buy_ 100 token B for token A, the swap would look like:
```
Swap { from: B, to: A, amount: 100 }
```
Which gives an amount in token A (our sell token) which can be swapped for 100 of token B (our buy token). It also gives a fee in token A. $\Box$

So, this `Swap` abstraction does exactly what we want (estimate the token price if we want to swap a specific amount), while being quite a bit simpler than the `Order` abstraction used by the protocol. The slight drawback is that the client has to convert the fee into the buy token themselves. Alternatively, the prices service could easily return the fee in _both_ tokens, so the client doesn't have to do anything.

Reply:
> The prices service could easily return the fee in both tokens, so the client doesn't have to do anything.

This is probably the way to go, actually.

Reply:
In principle this sounds good, but it will result us undercharging fees for illiquid tokens.